Analytics: Refactor Checkout page view tracking

Copons opened the pull request:
Fix #23508

Routes to test:

- `/checkout/features/:feature/:site/:plan`
- `/checkout/features/:feature/:site`
- `/checkout/:product/renew/:purchaseId/:site`
- `/checkout/:site/:product`
- `/checkout/:site`
- `/checkout/:site/with-gsuite/:domain/:receiptId`
- `/checkout/:site/with-gsuite/:domain`
- `/checkout/thank-you/no-site/pending/:orderId`
- `/checkout/thank-you/:site/pending/:orderId`
- `/checkout/thank-you/no-site/:receiptId`
- `/checkout/thank-you/no-site`
- `/checkout/thank-you/:site/:receiptId`
- `/checkout/thank-you/:site`
- `/checkout/thank-you/:site/:receiptId/with-gsuite/:gsuiteReceiptId`
- `/checkout/thank-you/:site/:receiptId/with-gsuite`
- `/checkout/thank-you/features/:feature/:site/:receiptId`
- `/checkout/thank-you/features/:feature/:site`

### Notes

This PR removes all `sectionifyWithRoutes` calls.
It would be good as a follow-up to remove it altogether.

## Testing instructions

- Enable Tracks debugging on Calypso by executing this in the developer console:
`localStorage.setItem('debug', 'calypso:analytics:tracks')`.
- Filter the console output with `Recording event "calypso_page_view" with actual props` to make it easier to read.
- Navigate all the routes above and make sure that all of them fire a `calypso_page_view` track event.

see also p5XAZ9-1J0-p2

Copons commented:
@markryall pinging you for your work in #23027.

Do you have any suggestions on how to test all these `/checkout` routes?

I can't seem to find any records for `/checkout/features/*` or `/checkout/thank-you/features/*` paths in our analytics db (querying for the last few months).
Might it be that those are obsolete paths that can be removed altogether?

Also and I can't seem to get into a `*/with-gsuite/*` path when trying to purchase a plan with GSuite. Is there a trick there that you know of?

yanirs commented:
> In other words, the page correctly records a /checkout/:site/, and the subcomponent also records a /checkout/domain-registration-details page view, which, though, doesn't match the path.
> My first idea was to remove it, but then: do you have any use for that, or it's just something that might clutter your analysis and I can safely remove it?

@Copons that's strange... I haven't seen this path before so I have no objections to removing it, but it's probably worth checking that it's not used in any active funnels.

Copons commented:
> that's strange... I haven't seen this path before so I have no objections to removing it, but it's probably worth checking that it's not used in any active funnels.

Good call @yanirs, I've found it (also notice that I mistyped it: it's actually `/checkout/domain-contact-information`) in several funnels that, at a quick glance, are 1+ years old.
I'd argue that it's an incorrect use of a page view event, but I'll check with the domains folks to see if those funnels are still relevant.

gwwar commented:
@Copons any blockers on this one? We could ask for help from nux.

Copons commented:
@gwwar Only blocker is that I can't figure out how to test most paths.
I've already pinged Amber and @Automattic/neutron in my audit post (p7jreA-1Cu-p2), and also on Slack, IIRC.
Though, if you know more appropriate teams for this, I'm open to new pings! 🙂

gwwar commented:
I'll see if I can wrangle getting some help. If we mainly need to test the listed routes, I think removing the in progress label and adding needs review is more appropriate.

Copons commented:
> I'll see if I can wrangle getting some help. If we mainly need to test the listed routes, I think removing the in progress label and adding needs review is more appropriate.

@gwwar Thank you. 🙇 
I've kept this WIP because I didn't want to submit a possibly broken PR with changes spanning across an entire section.
Though, I guess the Needs Review label would indeed be more helpful for getting reviews.

gwwar commented:
Comment reserved for testing. Let's make our best effort here, and if we can't find a particular route, that's okay, we can probably 🚢 and iterate later if we get more confusing data.

- [ ] `/checkout/features/:feature/:site/:plan`
- [ ] `/checkout/features/:feature/:site`
- [ ] `/checkout/:product/renew/:purchaseId/:site`
- [ ] `/checkout/:site/:product`
- [ ] `/checkout/:site`
- [ ] `/checkout/:site/with-gsuite/:domain/:receiptId`
- [ ] `/checkout/:site/with-gsuite/:domain`
- [ ] `/checkout/thank-you/no-site/pending/:orderId`
- [ ] `/checkout/thank-you/:site/pending/:orderId`
- [ ] `/checkout/thank-you/no-site/:receiptId`
- [ ] `/checkout/thank-you/no-site`
- [ ] `/checkout/thank-you/:site/:receiptId`
- [ ] `/checkout/thank-you/:site`
- [ ] `/checkout/thank-you/:site/:receiptId/with-gsuite/:gsuiteReceiptId`
- [ ] `/checkout/thank-you/:site/:receiptId/with-gsuite`
- [ ] `/checkout/thank-you/features/:feature/:site/:receiptId`
- [ ] `/checkout/thank-you/features/:feature/:site`

gwwar commented:
@Copons maybe verify that a real gsuite purchase looks okay (it's an add on when we create a new site). the`onboarding-checklist` is on in production so I think we won't really hit most thank you pages in existing flows.

Once this is done I think this looks mostly okay to 🚢 

It does look like we need to do some tidying with regards to flows and thank you pages, some of these can probably be retired and removed from the code base.

Copons commented:
@gwwar Did some additional testing on ❌ paths.

- `/checkout/features/:feature/:site/:plan`

Do we even use the `/features` paths? 🤔 

- `/checkout/thank-you/no-site/:receiptId`
- `/checkout/thank-you/no-site/pending/:orderId`

These two don't have a site fragment but `getSiteFragment` will think that `:receiptId` is a site ID, and, since it doesn't match the selected site ID, `PageViewTracker` won't record the event.

- `/checkout/thank-you/:site/pending/:orderId`
- `/checkout/thank-you/:site/:receiptId/with-gsuite/:gsuiteReceiptId`
- `/checkout/thank-you/:site/:receiptId/with-gsuite`

While the site fragment is in a very cool place that would make our lives so much easier, it's... well... not really compliant to the fact that Calypso expects it in last or second-to-last position...
I've already [worked around this for `/me/purchases/` paths](https://github.com/Automattic/wp-calypso/blob/8e152219f208c879869124856c53246f2b06309e/client/lib/route/path.js#L31-L43), but I'm less eager to do the same for what look like a soon-to-be-deprecated view.

We're currently [A/B testing](https://github.com/Automattic/wp-calypso/blob/8e152219f208c879869124856c53246f2b06309e/client/lib/abtest/active-tests.js#L68-L85) between `checklist` and `thank-you`, so it means that at least some users will see the `thank-you` page.

Though, these tests have been going on for more than 7 months now, and there's been something weird going on recently (p8Eqe3-nr-p2, p91bmw-sX-p2, cc @fditrapani @anneforbush).
Imho for now we should avoid adding more workarounds to the `path.js` lib, and only do it if `checklist` doesn't test well and we revert to `thank-you`.

gwwar commented:
@copons to unblock you, let's ask if we plan on having these ❌ routes still be active. If so, I think we'll need to add an exception for this and also PSA on crafting routes, since Calypso does make a lot of assumptions.

Copons commented:
@gwwar quick update: the checklist is not yet ready for prime time, and so we should treat the thank you pages as the de facto standard for the time being.
I'm going to update this PR by whitelisting them in `PageViewTracker`.

Copons commented:
@gwwar to recap:

In fb93ba5 I've added the `thank-you` exceptions in `getSiteFragment`.

In 316c477 I've addressed an issue that coincidentally popped up when testing a `thank-you` route: Tracks events properties must be in snake_case.
In my test, the page view was rejected because I was sending `receiptId` instead of `receipt_id`.
So, I've replaced all camel case page view properties names in the `/checkout` folder with their snake case equivalent.

I think this is pretty much ready to go now, but I'd appreciate a last look. 🙇

gwwar commented on client/lib/route/path.js:
```diff
@@ -28,10 +28,17 @@ export function getSiteFragment( path ) {
-	// Though, in some`/me/purchases` paths, it could also be in third position,
-	// right after the `/me/purchases/` part.
+	// Avoid confusing the receipt ID for the site ID in domain-only checkouts.
+	if ( 0 === basePath.indexOf( '/checkout/thank-you/no-site/' ) ) {
```
I don't think we're super strict about this, but we don't really use yoda conditions in our JS code.

Copons replied:
https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#yoda-conditions

Not enforced, but also welcomed! 😄

gwwar replied:
😆 I will continue to ignore this suggestion then, until we autoformat this too!

gwwar commented on client/my-sites/checkout/checkout-thank-you/index.jsx:
```diff
@@ -267,13 +267,13 @@ export class CheckoutThankYou extends React.Component {
 			return {
 				path: '/checkout/thank-you/:site/:receipt_id/with-gsuite/:gsuite_receipt_id',
-				properties: { gsuiteReceiptId, receiptId, site },
+				properties: { gsuite_receipt_id: gsuiteReceiptId, receipt_id: receiptId, site },
```
Good catch on the 🐍 case

gwwar commented:
Overall looks good! Thanks @Copons ✨  Feel free to 🚢 after you have some time to 👀 landing changes in prod.

I couldn't repro one route, but I'm pretty sure it's not being actively used. Remind me to file a few bugs too. The thank-you pages don't render correctly with a clean browser cache (eg browser persistence turned off).
